From 049ab0dd94544bb0770defa54b1322812e6d3082 Mon Sep 17 00:00:00 2001
From: Dennis Gilmore <dennis@ausil.us>
Date: Mon, 4 Aug 2014 05:02:08 -0500
Subject: [PATCH] correctly call pykickstarts makeVersion

---
 builder/kojid | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/builder/kojid b/builder/kojid
index b23e9ce..2451ea6 100755
--- a/builder/kojid
+++ b/builder/kojid
@@ -2387,9 +2387,9 @@ class ImageTask(BaseTaskHandler):
         #      may not be a problem if the included kickstarts are present
         #      in the repository we checked out.
         if opts.get('ksversion'):
-            version = ksparser.makeVersion(ksparser.stringToVersion(opts['ksversion']))
+            version = ksparser.version.makeVersion(ksparser.stringToVersion(opts['ksversion']))
         else:
-            version = ksparser.makeVersion()
+            version = ksparser.version.makeVersion()
         self.ks = ksparser.KickstartParser(version)
         try:
             self.ks.readKickstart(kspath)
@@ -2768,10 +2768,10 @@ class OzImageTask(BaseTaskHandler):
         #      macros, Oz will fail because it can only handle flat files.
         #      We require users to flatten their kickstart file.
         if self.opts.get('ksversion'):
-            version = ksparser.makeVersion(
+            version = ksparser.version.makeVersion(
                 ksparser.stringToVersion(self.opts['ksversion']))
         else:
-            version = ksparser.makeVersion()
+            version = ksparser.version.makeVersion()
         ks = ksparser.KickstartParser(version)
         self.logger.debug('attempting to read kickstart: %s' % kspath)
         try:
-- 
2.0.4

