From 073bdb69e73e8943f3fb031f56b3d3ca66412edc Mon Sep 17 00:00:00 2001
From: Jay Greguske <jgregusk@redhat.com>
Date: Tue, 15 Apr 2014 14:45:40 -0400
Subject: [PATCH 2/8] compress docker tarball properly

---
 builder/kojid | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/builder/kojid b/builder/kojid
index 1691a54..cee7637 100755
--- a/builder/kojid
+++ b/builder/kojid
@@ -2824,7 +2824,8 @@ class BaseImageTask(OzImageTask):
 
         # target-image type images
         if 'docker' in self.formats:
-            targ = do_target_image(base.base_image.identifier, 'docker')
+            targ = do_target_image(base.base_image.identifier, 'docker',
+                ova_opts={'compress': 'gzip'})
             images['docker'] = {'image': targ.target_image.data}
 
         ova_opts = {}
-- 
2.0.4

