From 7a1d6c4cd42f6efe8a3833a77a4475b6cfe9da43 Mon Sep 17 00:00:00 2001
From: Jay Greguske <jgregusk@redhat.com>
Date: Wed, 16 Apr 2014 12:00:33 -0400
Subject: [PATCH 4/8] define bld_info before the try in case of an earlier
 exception

---
 builder/kojid | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/builder/kojid b/builder/kojid
index 32400c6..fa5daca 100755
--- a/builder/kojid
+++ b/builder/kojid
@@ -1824,6 +1824,7 @@ class BuildBaseImageTask(BuildImageTask):
             raise koji.ApplianceError, 'ImageFactory functions not available'
 
         # build image(s)
+        bld_info = None
         try:
             release = opts.get('release')
             if not release:
@@ -1832,7 +1833,6 @@ class BuildBaseImageTask(BuildImageTask):
                 raise koji.ApplianceError('The Version may not have a hyphen')
             if '-' in release:
                 raise koji.ApplianceError('The Release may not have a hyphen')
-            bld_info = None
             if not opts.get('scratch'):
                 bld_info = self.initImageBuild(name, version, release,
                     target_info, opts)
-- 
2.0.4

