From 61c1737652544150fe962921b9dfd3708e1d256e Mon Sep 17 00:00:00 2001
From: Jay Greguske <jgregusk@redhat.com>
Date: Thu, 17 Apr 2014 18:59:41 -0400
Subject: [PATCH 6/8] use offline icicle generation

---
 builder/kojid | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/builder/kojid b/builder/kojid
index fa5daca..4712cc5 100755
--- a/builder/kojid
+++ b/builder/kojid
@@ -2787,7 +2787,8 @@ class BaseImageTask(OzImageTask):
         tlog.setLevel(logging.DEBUG)
         tlog.addHandler(fhandler)
         images = {}
-        params = {'install_script': str(self.ks.handler)} #ks contents
+        params = {'install_script': str(self.ks.handler),
+                  'offline_icicle': True}
         random.seed() # necessary to ensure a unique mac address
         try:
             base = bd.builder_for_base_image(template, parameters=params)
-- 
2.0.4

