From 1fbbf825ab7962e68d0059ab86b0955312997035 Mon Sep 17 00:00:00 2001
From: Jay Greguske <jgregusk@redhat.com>
Date: Thu, 8 May 2014 11:34:51 -0400
Subject: [PATCH 8/8] bump install timeout to 2 hours

---
 builder/kojid | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/builder/kojid b/builder/kojid
index 4712cc5..90ff2bd 100755
--- a/builder/kojid
+++ b/builder/kojid
@@ -2608,7 +2608,7 @@ class OzImageTask(BaseTaskHandler):
             'imgdir': os.path.join(self.workdir, 'scratch_images'),
             'tmpdir': os.path.join(self.workdir, 'oz-tmp'),
             'verbose' : True,
-            'timeout': 3600,
+            'timeout': 7200,
             'output': 'log',
             'raw': False,
             'debug': True,
-- 
2.0.4

