diff -uNr koji-1.10.1.orig/cli/koji.conf koji-1.10.1/cli/koji.conf
--- koji-1.10.1.orig/cli/koji.conf	2016-03-01 16:43:15.000000000 -0600
+++ koji-1.10.1/cli/koji.conf	2016-03-01 16:57:55.875274789 -0600
@@ -3,26 +3,23 @@
 ;configuration for koji cli tool
 
 ;url of XMLRPC server
-;server = http://hub.example.com/kojihub
+server = http://koji.fedoraproject.org/kojihub
 
 ;url of web interface
-;weburl = http://www.example.com/koji
+weburl = http://koji.fedoraproject.org/koji
 
 ;url of package download site
-;pkgurl = http://www.example.com/packages
+topurl = https://kojipkgs.fedoraproject.org/
 
 ;path to the koji top directory
 ;topdir = /mnt/koji
 
-;configuration for Kerberos authentication
-
-;the service name of the principal being used by the hub
-;krbservice = host
+anon_retry = true
 
 ;configuration for SSL authentication
 
 ;client certificate
-;cert = ~/.koji/client.crt
+cert = ~/.fedora.cert
 
 ;certificate of the CA that issued the HTTP server certificate
-;serverca = ~/.koji/serverca.crt
+serverca = ~/.fedora-server-ca.cert
